go-apps/meep-tc-engine/ip-mgr.go

```diff
@@ -0,0 +1,244 @@
+		}
+	}
+}
+
+// SetSvcList - Set list of services to monitor IP addresses for
+func (im *IpManager) SetSvcList(svcList map[string]bool) {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	// Remove stale entries
+	for svcName := range im.svcIpMap {
+		if _, found := svcList[svcName]; !found {
+			delete(im.svcIpMap, svcName)
+		}
+	}
+
+	// Add missing entries
+	for svcName := range svcList {
+		if _, found := im.svcIpMap[svcName]; !found {
+			im.svcIpMap[svcName] = IP_ADDR_NONE
+		}
+	}
+}
+
+// GetPodIp - Get Pod IP address
+func (im *IpManager) GetPodIp(podName string) string {
+	podIp, found := im.podIpMap[podName]
+	if !found {
+		podIp = IP_ADDR_NONE
+	}
+	return podIp
+}
+
+// GetPodIp - Get Svc IP address
+func (im *IpManager) GetSvcIp(svcName string) string {
+	svcIp, found := im.svcIpMap[svcName]
+	if !found {
+		svcIp = IP_ADDR_NONE
+	}
+	return svcIp
+}
+
+// Start - Start monitoring IP addresses
+func (im *IpManager) Start() error {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	log.Debug("Starting IP Address Manager: ", im.name)
+
+	// Make sure ticker is not already running
+	if im.ticker != nil {
+		return errors.New("Ticker already running")
+	}
+
+	// Start ticker to periodically retrieve platform information
+	im.ticker = time.NewTicker(DEFAULT_TICKER_INTERVAL_MS * time.Millisecond)
+	go func() {
+		for range im.ticker.C {
+			im.mutex.Lock()
+
+			// Refresh IP addresses
+			im.refreshIpAddresses()
+
+			im.mutex.Unlock()
+		}
+	}()
+
+	return nil
+}
+
+// Stop - Stop monitoring IP addresses
+func (im *IpManager) Stop() {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	if im.ticker != nil {
+		im.ticker.Stop()
+		im.ticker = nil
+	}
+}
+
+// Refresh - Request a IP addresses
+func (im *IpManager) Refresh() {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	im.refreshIpAddresses()
+}
+
+// refreshIpAddresses - Refresh Pod & Service IP addresses
+func (im *IpManager) refreshIpAddresses() {
+	var err error
+	updated := false
+
+	// Establish connection to K8s API Server if not connected
+	if !im.isConnected {
+		im.clientset, err = connectToAPISvr()
+		if err != nil {
+			log.Error("Failed to connect with k8s API Server. Error: ", err)
+			return
+		}
+		im.isConnected = true
+	}
+
+	// Retrieve all sandbox pods from k8s api
+	podList, err := im.clientset.CoreV1().Pods(im.sandboxName).List(metav1.ListOptions{})
+	if err != nil {
+		log.Error("Failed to retrieve pods from k8s API Server. Error: ", err)
+		im.isConnected = false
+		return
+	}
+
+	// Refresh Pod IPs
+	for _, pod := range podList.Items {
+		podName := pod.ObjectMeta.Labels["meepApp"]
+		podIp := pod.Status.PodIP
+
+		if ip, found := im.podIpMap[podName]; found && podIp != "" && podIp != ip {
+			log.Debug("Setting podName: ", podName, " ip: ", podIp)
+			im.podIpMap[podName] = podIp
+			updated = true
+		}
+	}
+
+	// Retrieve all sandbox services from k8s api
+	svcList, err := im.clientset.CoreV1().Services(im.sandboxName).List(metav1.ListOptions{})
```

Comment:
I believe the above query will return all pods in the sandbox (core & scenario)
e.g. `kubectl get service -n <sandboxName>`

For efficiency, it should only focus at the scenario pods
e.g ` kubectl get service -n <sandboxName> -l meepOrigin=scenario`

Unfortunately, the services do not have the meepOrigin label - make sure that scenario created services get the same label set as the pods

Reply:
The meep-webhook does not inject the meepOrigin label on services, therefore it is not guaranteed that it will be present on all scenario services (e.g. user chart services). The finest granularity for services is then the sandbox/namespace, which is what is used in the code. To avoid modifying the meep-webhook at this point, I recommend keeping the code as is and getting the IP addresses for all sandbox services.

Additional note: meep-webhook requires the user-defined helm charts to set the 'release' label on its deployments, which allows the webhook to identify scenario pods where the meep-tc-sidecar must be injected. Typical helm charts set this label by default on deployments, however they do not set them on services. Adding support for service label injection in the meep-webhook would add a requirement on user-charts to set the 'release' label on all services as well. To reduce complexity and avoid this additional requirement it makes sense to add slightly more processing in the IP Manager code to retrieve IP addresses for all sandbox services instead of only the scenario services.